lib/builder-webpack5/src/preview/iframe-webpack.config.ts

```diff
@@ -9,7 +9,7 @@ import TerserWebpackPlugin from 'terser-webpack-plugin';
 import VirtualModulePlugin from 'webpack-virtual-modules';
 import ForkTsCheckerWebpackPlugin from 'fork-ts-checker-webpack-plugin';
 
-import type { Options, CoreConfig } from '@storybook/core-common';
+import type { Options, CoreConfig, Webpack5BuilderConfig } from '@storybook/core-common';
```

Comment:
This should be cleaned up. There shouldn't be anything in `core-common` that refers to webpack. We can do it in a separate PR

Reply:
@shilman the entry point that this PR touches (`to-importFn` etc) is webpack specific, but common to WP4 + 5. Should we also move that somewhere?

Reply:
If this is a 7.0 change it should be in `core-webpack`. If it's a 6.x change, I guess it can stay 🤷

Reply:
Happy to make that refactor in 7.0 otherwise though.

Reply:
Let's make the refactor here, release in 7.0, and then potentially backport after we've tested. WDYT?

Reply:
@shilman I moved the utilities over. I don't know how to move the `Webpack5BuilderConfig` type while keeping:

```
export type BuilderConfig =
  | BuilderName
  | BuilderConfigObject
  | Webpack4BuilderConfig
  | Webpack5BuilderConfig;
```

Reply:
How about something like the following.

In `core-common`:

```
export type BuilderConfig =
  | BuilderName
  | BuilderConfigObject
```

In `core-webpack` (or `builder-webpack5`?):

```
import type { BuilderConfig as BaseBuilderConfig } from '@storybook/core-common';

export type BuilderConfig = BaseBuilderConfig | Webpack5BuilderConfig;
```

Reply:
Well but the problem is that `BuilderConfig` is part of `CoreConfig` which is part of `StorybookConfig` which the user imports when defining their SB. 

```ts
import type { StorybookConfig } from '@storybook/react-webpack5/types`
```

Probably what we should do is parameterize it?

```ts
// In `core-common`
type StorybookConfig<TBuilderConfig = BuilderConfig>

// In `react-webpack5`
import type { StorybookConfig as BaseStorybookConfig } from '@storybook/core-common';
import type { BuilderConfig } from '@storybook/builder-webpack5';

export type StorybookConfig = BaseStorybookConfig<BuilderConfig>;
```

Does that sounds right? WDYT @ndelangen?

Reply:
Generally the idea is to keep composing the StorybookConfig up into the framework.

https://github.com/storybookjs/storybook/blob/ea5c7a7616fb7f22b13315a3a16f6a8484864583/frameworks/react-webpack5/src/types.ts#L47-L49

https://github.com/storybookjs/storybook/blob/ea5c7a7616fb7f22b13315a3a16f6a8484864583/presets/react-webpack/src/types.ts#L51-L55

https://github.com/storybookjs/storybook/blob/8e84f36d0f3abc75867b6b69de3e1c12c23b9166/lib/core-webpack/src/types.ts#L41-L42

---

But -correct me if I'm wrong- but we expect users to only have to specify the builder config if they are NOT using a framework, right?
I think using a framework AND setting a builder should not really happen?

Placing the builder config 

---

So if the above is true, the main question is: where does the user import the "StorybookConfig" from if they are not using a framework?

I think the answer is: probably the renderer.

But alternatively we ask the user to do the following:

```ts
// user's main.js
import type { StorybookConfig as StorybookConfigBase } from '@storybook/react`;
import type { StorybookConfig as StorybookConfigBuilder } from '@storybook/builder-webpack5`;

type StorybookConfig = StorybookConfigBase & StorybookConfigBuilder


module.exports: StorybookConfig = {
  // ...
};
```

We could do the above in the framework, move this code:
https://github.com/storybookjs/storybook/blob/414d530df317b3b187d4df76380dba9bb27209c1/lib/core-common/src/types.ts#L18-L24
into `builder-webpack5` package, import it from the framework and compose the types in the framework package. That seems to be the correct thing to do.

1 thing is extremely clear though: we should not get into the position where we have cyclical dependencies in our types.

Reply:
@ndelangen 

> But -correct me if I'm wrong- but we expect users to only have to specify the builder config if they are NOT using a framework, right?
> I think using a framework AND setting a builder should not really happen?

In this case we are talking about the builder options like `lazyCompilation` or `fsCache`. So I don't think your assumption here is quite right.

Reply:
I think those should be framework options, flowing into the builderOptions then.

I think we want to move these builderOptions into the builder, and adjust the frameworkOptions to take them, and flow them into the builder.

Reply:
@shilman are you OK with the current state of this PR assuming the above gets sorted in https://linear.app/chromaui/issue/SB-347/sequence-importing-stories-in-parallel-in-webpack5-builder